northstar/src/runtime/repository.rs

```diff
@@ -12,100 +12,305 @@
+                    let container = Container::new(name, version);
+                    Result::<_, Error>::Ok((container, file, npk))
+                })
+                .then(|r| match r {
+                    Ok(r) => ready(r),
+                    Err(_) => panic!("Task error"),
+                });
+                loads.push(task);
+            } else {
+                debug!("Skipping {}", file.display());
             }
+        }
 
-            let npk = task::block_in_place(|| Npk::from_path(entry.path().as_path(), None))
-                .map_err(Error::Npk)?;
-            let name = npk.manifest().name.clone();
-            let version = npk.manifest().version.clone();
-            let container = Container::new(name, version);
-            containers.insert(container.clone(), (entry.path(), npk.manifest().clone()));
+        let results = join_all(loads).await;
+        for result in results {
+            match result {
+                Ok((container, file, npk)) => {
+                    containers.insert(container, (file, Arc::new(npk)));
+                }
+                Err(e) => warn!("Failed to load: {}", e),
+            }
         }
 
-        Ok(Repository {
+        let duration = start.elapsed();
+        info!(
+            "Loaded {} containers from {} in {:.03}s (avg: {:.05}s)",
+            containers.len(),
+            dir.display(),
+            duration.as_fractional_secs(),
+            duration.as_fractional_secs() / containers.len() as f64
+        );
+
+        Ok(DirRepository {
             id,
             dir,
-            key: key.await.transpose().map_err(Error::Key)?,
+            key,
             containers,
         })
     }
+}
 
-    pub async fn add(&mut self, container: &Container, src: &Path) -> Result<(), Error> {
-        let dest = self
-            .dir
-            .join(format!("{}-{}.npk", container.name(), container.version()));
-
-        // Check if the npk already in the repository
-        if dest.exists() {
-            return Err(Error::InstallDuplicate(container.clone()));
+#[async_trait::async_trait]
+impl<'a> Repository for DirRepository {
+    async fn insert(&mut self, rx: &mut Receiver<Bytes>) -> Result<Container, Error>
+    where
+        Self: Sized,
+    {
+        let dest = self.dir.join(format!("{}.npk", uuid::Uuid::new_v4()));
+        let mut file = fs::File::create(&dest)
+            .await
+            .map_err(|e| Error::Io("Failed create npk in repository".into(), e))?;
+        while let Some(r) = rx.recv().await {
+            file.write_all(&r)
+                .await
+                .map_err(|e| Error::Io("Failed to write npk".into(), e))?;
         }
-
-        // Copy the npk to the repository
-        fs::copy(src, &dest)
+        file.flush()
             .await
-            .map_err(|e| Error::Io("Failed to copy npk to repository".into(), e))?;
+            .map_err(|e| Error::Io("Failed to flush npk".into(), e))?;
+        drop(file);
```

Comment:
why do we need the drop here?

Reply:
The drop is not needed explicitly because the `flush` should do the trick. Anyways I wanted to get the fd to be closed before the npk is opened again for parsing.